File: backend/src/services/template.service.js
import prisma from '../prisma/client.js';

class TemplateService {
  async createProjectTemplate({ name, description, workspaceId, createdBy, tasks }) {
    return prisma.projectTemplate.create({
      data: {
        name,
        description,
        workspaceId,
        createdBy,
        tasks: {
          create: tasks,
        },
      },
      include: {
        tasks: true,
      },
    });
  }

  async getProjectTemplates(workspaceId) {
    return prisma.projectTemplate.findMany({
      where: { workspaceId },
      include: {
        tasks: true,
      },
    });
  }

  async getProjectTemplateById(templateId) {
    return prisma.projectTemplate.findUnique({
      where: { id: templateId },
      include: {
        tasks: true,
      },
    });
  }

  async applyProjectTemplate(templateId, { name, workspaceId, managerId }) {
    const projectTemplate = await this.getProjectTemplateById(templateId);
    if (!projectTemplate) {
      throw new Error('Project template not found');
    }

    const project = await prisma.project.create({
      data: {
        name,
        description: projectTemplate.description,
        workspaceId,
        managerId,
        tasks: {
          create: projectTemplate.tasks.map(task => ({
            title: task.title,
            description: task.description,
            category: task.category,
            estimatedHours: task.estimatedHours,
          })),
        },
      },
      include: {
        tasks: true,
      },
    });

    return project;
  }
}

export default new TemplateService();

File: creapolis_app/lib/data/services/template_service.dart
import 'package:dio/dio.dart';
import 'package:injectable/injectable.dart';
import '../../../core/services/api_service.dart';
import '../../../domain/entities/project_template.dart';

@injectable
class TemplateService {
  final ApiService _apiService;

  TemplateService(this._apiService);

  Future<List<ProjectTemplate>> getProjectTemplates(int workspaceId) async {
    try {
      final response = await _apiService.get('/templates/project-templates/$workspaceId');
      return (response.data as List)
          .map((json) => ProjectTemplate.fromJson(json))
          .toList();
    } on DioError catch (e) {
      // Handle error
      return [];
    }
  }

  Future<ProjectTemplate?> createProjectTemplate({
    required String name,
    String? description,
    required int workspaceId,
    required List<Map<String, dynamic>> tasks,
  }) async {
    try {
      final response = await _apiService.post(
        '/templates/project-templates/$workspaceId',
        data: {
          'name': name,
          'description': description,
          'workspaceId': workspaceId,
          'tasks': tasks,
        },
      );
      return ProjectTemplate.fromJson(response.data);
    } on DioError catch (e) {
      // Handle error
      return null;
    }
  }

  Future<Project?> applyProjectTemplate({
    required int templateId,
    required String projectName,
    required int workspaceId,
    int? managerId,
  }) async {
    try {
      final response = await _apiService.post(
        '/templates/project-templates/$templateId/apply',
        data: {
          'name': projectName,
          'workspaceId': workspaceId,
          'managerId': managerId,
        },
      );
      // Assuming the backend returns the created project
      return Project.fromJson(response.data);
    } on DioError catch (e) {
      // Handle error
      return null;
    }
  }
}
